### PR TITLE
OpInfo tests for `nn.functional.max_pool{n}d`.

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. 

It is planned to use these tests for fixing issues with the max_unpools' backward methods reported in https://github.com/pytorch/pytorch/issues/67658 and https://github.com/pytorch/pytorch/issues/67657.
max_unpool.backward methods are not tested and implemented with custom kernels. We can replace these kernels with advanced indexing operations (i.e. `gather`) which are efficient and well tested.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -9445,14 +9486,42 @@ def ref_pairwise_distance(input1, input2):
...
+           aten_name='max_pool3d',
+           supports_autograd=True,
+           supports_out=False,
+           assert_jit_shape_analysis=True,
+           dtypesIfCPU=floating_types(),
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+           # TODO: investigate nondeterminism
+           gradcheck_nondet_tol=GRADCHECK_NONDET_TOL,
+           decorators=(
+               # Fails JIT output shape checks
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9521]: cc @eellison

**Review by mruberry — APPROVED** [review]
Hey @nikitaved! This looks great!

I have one question whether a `detach()` should be added in one place.

Let's see what the extended tests say, too.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -3087,38 +3087,79 @@ def generator():
 
     return list(generator())
 
+class _TestParamsMaxPool1d(object):
+
+    def __init__(self):
+        self.kwargs = {
+            'kernel_size': [(2,), 3],
+            'stride': [2],
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:3095]: Do you think testing `stride=None` will be useful as well?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -3087,38 +3087,79 @@ def generator():
...
+            'ceil_mode': [True, False],
+            'padding': [0, 1],
+            'dilation': [1],
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:3098]: (This applies for other args as well) - Do you think checking these values as tuples is useful? Something like: `(1,)` and `1`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -9445,14 +9486,42 @@ def ref_pairwise_distance(input1, input2):
...
            assert_jit_shape_analysis=True,
            dtypesIfCPU=floating_types(),
            dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
-           sample_inputs_func=sample_inputs_max_pool2d),
+           sample_inputs_func=sample_inputs_max_pool),
+    OpInfo('nn.functional.max_pool3d',
+           aten_name='max_pool3d',
+           supports_autograd=True,
+           supports_out=False,
+           assert_jit_shape_analysis=True,
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:9515]: This is interesting, was it added recently? From what I remember, this test was failing the last time I made the PR (https://github.com/pytorch/pytorch/pull/67005/files#diff-18a0c3e7cd07089b34760028c387d39de1aa393172f1d5f123692106b3a3d144R9461)

**Review by krshrimali — APPROVED** [review]
Hi, @nikitaved - Thanks for the PR, looks nice - I added a few comments for you to take a look at, this LGTM (approving it since the comments are mostly minor suggestions) but I'm looking forward to your comments. :)

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -3087,38 +3087,79 @@ def generator():
...
-                "stride": max_pool.stride,
-                "padding": max_pool.padding,
-                "dilation": max_pool.dilation,
-                "ceil_mode": max_pool.ceil_mode,
-                "return_indices": max_pool.return_indices,
-            }
-            sample_input = make_arg((N, C, H, W)) if N is not None else (make_arg((C, H, W)))
+        params_generator = params_generator_type_dict[op_info.name]()
+        for (shape, memory_format), kwargs in params_generator.gen_input_params():
+            arg = make_arg(shape).to(memory_format=memory_format).requires_grad_(requires_grad)
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:3160]: No need for `detach` because `make_arg` creates tensors with `requires_grad=False`.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:3098]: I agree!

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:9515]: Do you know how I could fix this shape analysis. I did some light debugging and the shape analysis test was matching super wrong output shapes.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:3095]: Makes sense!

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:9515]: Anyway, what is the strategy with OpInfo testing? Do we try to enable them all and then deprecate whatever is failing, or do we disable failing tests with flags? I personally find the former way a bit more future proof. What do you think, @mruberry ?

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:9515]: That's a good question, Nik! The thing with shape analysis tests is, that they aren't really implemented yet for these ops. (`max_pool2d` has the implementation though).

Since this is more like an absence of implementation, and not necessarily a failure, hence disabling these tests might be a better option. (IMO)

Please see: https://github.com/pytorch/pytorch/pull/67005#discussion_r733726615 for @eellison's comment.

If you are curious, here is where the shape analysis was added for max_pool2d: https://github.com/pytorch/pytorch/pull/63530

Please let me know if you have any follow up questions.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:9515]: cc @eellison 

If a test is failing and the default is not to test it then I'd stick with that instead of enabling the test and then adding an xFail or skip decorator for it.

More generally if a test is controlled through metadata then I'd prefer using the metadata rather than decorators to skip it.

**Comment by nikitaved** [comment]
@krshrimali , @mruberry , do you think it makes sense to add OpInfo entries for `_pool_with_indices` here as well? Or shall we leave it to https://github.com/pytorch/pytorch/pull/67005?

**Comment by mruberry** [comment]
> @krshrimali , @mruberry , do you think it makes sense to add OpInfo entries for `_pool_with_indices` here as well? Or shall we leave it to #67005?

I think we can leave them as a follow-up PR

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -9515,14 +9566,40 @@ def ref_pairwise_distance(input1, input2):
...
+           check_batched_gradgrad=False,
            supports_out=False,
            assert_jit_shape_analysis=True,
            dtypesIfCPU=floating_types(),
            dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
-           sample_inputs_func=sample_inputs_max_pool2d),
+           decorators=(
+               # BUG: Fails JIT output shape checks when stride=None
+               DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
+           ),
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:9590]: @mruberry , shall we open an issue?

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:9590]: Done. https://github.com/pytorch/pytorch/issues/68117

Comment (eellison) [reply torch/testing/_internal/common_methods_invocations.py:9590]: Can you just remove the `assert_jit_shape_analysis` ? We're not giving incorrect answers we're just failing to propagate the shape

**Comment by eellison** [comment]
@nikitaved do you mind making the following change and removing the jit skip ? looks like a bug in ternary processing from the C++ parser.
```
diff --git a/torch/csrc/jit/runtime/symbolic_shape_registry.cpp b/torch/csrc/jit/runtime/symbolic_shape_registry.cpp
index 3df81ea400..6dae187c55 100644
--- a/torch/csrc/jit/runtime/symbolic_shape_registry.cpp
+++ b/torch/csrc/jit/runtime/symbolic_shape_registry.cpp
@@ -181,7 +181,12 @@ const std::string shape_compute_functions =
 
           assert len(stride) == 0 or len(stride) == 1 or len(stride) == 2, "max_pool2d: stride must either be omitted, a single int, or a tuple of two ints"
           dH = kH if len(stride) == 0 else stride[0]
-          dW = kW if len(stride) == 0 else dH if len(stride) == 1 else stride[1]
+          if len(stride) == 0:
+            dW = kW
+          elif len(stride) == 1:
+            dW = dH
+          else:
+            dW = stride[1]
 ```

**Comment by nikitaved** [comment]
Thank you, @eellison , this seems to have resolved the issue.

**Comment by nikitaved** [comment]
All issues are resolved. Unless there is something else to fix, this one seems ready to go.
